server/src/main/java/org/apache/druid/segment/realtime/firehose/TimedShutoffFirehoseFactory.java

```diff
@@ -63,31 +69,25 @@ public Firehose connect(InputRowParser parser, File temporaryDirectory) throws I
   class TimedShutoffFirehose implements Firehose
   {
     private final Firehose firehose;
-    private final ScheduledExecutorService exec;
-    private final Object shutdownLock = new Object();
-    private volatile boolean shutdown = false;
+    private final ScheduledExecutorService shutdownExec;
+    @GuardedBy("this")
+    private boolean closed = false;
 
     TimedShutoffFirehose(InputRowParser parser, File temporaryDirectory) throws IOException
     {
       firehose = delegateFactory.connect(parser, temporaryDirectory);
 
-      exec = Execs.scheduledSingleThreaded("timed-shutoff-firehose-%d");
-
-      exec.schedule(
-          new Runnable()
-          {
-            @Override
-            public void run()
-            {
-              log.info("Closing delegate firehose.");
-
-              shutdown = true;
-              try {
-                firehose.close();
-              }
-              catch (IOException e) {
-                log.warn(e, "Failed to close delegate firehose, ignoring.");
-              }
+      shutdownExec = Execs.scheduledSingleThreaded("timed-shutoff-firehose-%d");
```

Comment:
It seems to me we can close without creating an ExecutorService.

Reply:
Not sure what do you mean?